Fix GCI mounter issue

jingxu97 opened the pull request:
Two fixes in this PR

1. Remove break in the mounter script to make sure gc run multiple times
To make sure existed container created by mounter script gets garbage collected, we make rkt gc run multiple times

2. Fix unmount issue cased by GlusterFS mounter 
When GlusterFS volume is mounted on GCI image cluster, the GCI mounter script will create a container and then execute mount in the container. Because mount for GlusterFS will run a glusterfs daemon, the container created by the mounter will continue be in running state and rkt garbage collection will not collect it. Because of the shared mount propagation setting, any mounts on the host will be reflected to the container which might cause a large mount table. Also, in current umount device logic, we check the mount references before issuing the operations. In this case, unmount device will fail because of the extra mount points in the container. The following gives an example of original mount path and the shared mount point created in container.

a. original mount 
/var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/nfs-disk rw,relatime shared:133 - ext4 /dev/sdb rw,data=ordered
b. shared mount in container
/var/lib/rkt/pods/run/3496eadc-74d9-4349-9370-db7636f5f963/stage1/rootfs/opt/stage2/gci mounter/rootfs/var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/nfs-disk rw,relatime shared:133 - ext4 /dev/sdb rw,data=ordered

The current workaround for this issue is: since these shared mounts are irrelevant to the original mounts, they should be not considered when checking the mount references. By comparing the mount paths, the shared mount path contains the full original mount path so they should be filtered out.

jingxu97 committed:
Fix GCI mounter script to run garbage collection multiple times

Remove break in the mounter script to make sure gc run multiple times

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38124)
<!-- Reviewable:end -->

jingxu97 committed:
Fix unmount issue cuased by GCI mounter

this is a workaround for the unmount device issue caused by gci mounter. In GCI cluster, if gci mounter is used for mounting, the container started by mounter script will cause additional mounts created in the container. Since these mounts are irrelavant to the original mounts, they should be not considered when checking the mount references. By comparing the mount path prefix, those additional mounts can be filtered out.

Plan to work on better approach to solve this issue.

saad-ali commented on pkg/util/mount/mount.go:
```diff
@@ -130,20 +131,40 @@ func GetMountRefs(mounter Interface, mountPath string) ([]string, error) {
 			if mps[i].Device == deviceName && mps[i].Path != slTarget {
-				refs = append(refs, mps[i].Path)
+				mpPrefix := getPodsDirPrefix(mps[i].Path, options.DefaultKubeletPodsDirName)
+				if mpPrefix == podsDirPrefix {
```
My worry with this is that it may skip checks on non-rkt mounts as well. Let's see if we can minimize this.

jingxu97 replied:
Changed to filter out the mount refs based on string "gci-mounter" in the path

vishh commented on pkg/util/mount/mount.go:
```diff
@@ -130,14 +130,24 @@ func GetMountRefs(mounter Interface, mountPath string) ([]string, error) {
 		}
 	}
 
+	// TODO: this is a workaround for the unmount device issue caused by gci mounter.
```
This is a total hack. `gci_mounter` is a GKE specific deployment detail. We should not be leaking that logic into upstream.

vishh commented on cluster/gce/gci/mounter/mounter:
```diff
@@ -31,7 +31,7 @@ function gc {
     # Rkt pods end up creating new copies of mounts on the host. Hence it is ideal to clean them up right away.
     attempt=0
     until [ $attempt -ge 5 ]; do
-	${RKT_BINARY} gc --grace-period=0s &> /dev/null && break
```
Why are you making this change?

jingxu97 replied:
otherwise, it only run gc once.

vishh replied:
Why is that a problem?

jingxu97 replied:
when I test it, I think sometimes the container still in running state in the moment when gc is executed. Give a few try can give more chances to garbage collect the containers . I also thought it was intended to execute 5 times.

vishh replied:
It was only intended to run multiple times if any of the previous attempts fail. Blocking for 5 seconds doesn't sound like a good idea. Is it possible to detect incomplete gc and then retry?
Also why is GC of concern?

jingxu97 replied:
If the container is not GC, it causes those extra mounts which makes
1. mount table is very big if there are many mount points, which might be a performance concern since we list the mount points during unmount operations
2. Without this PR's fix, unmount device will fail because of those unrelated mount references. (umount device function first check whether there are other mount references for the mount path and will fail the operation if there are some exist)

vishh replied:
`1` might be of concern. Do we have any data to back up that concern?
I don't get `2`. Why is umount failing if other shared recursive mounts exist? Is the `umount` syscall failing or is it our own custom logic?

vishh replied:
@jingxu97 
> it is our own logic to check the mount references.

Can't we fix that logic then? As long as there is no separate mount point with the same device (bind mount), we can safely unmount. Check the mount reference number in `/proc/self/mountinfo` to make that decision.

vishh replied:
If a mount appears in two locations via a shared mount then `/proc/self/mountinfo` will explicitly indicate that. Here is an example.

```shell
$ cat /proc/self/mountinfo
...
90 80 0:66 / /tmp/original/tmpfs rw,relatime **shared:2** - tmpfs tmpfs rw
103 101 0:66 / /tmp/shared/tmpfs rw,relatime **shared:2** - tmpfs tmpfs rw
```

Look at the mount propagation type followed by a propagation number. If the number is the same, then that mount can be ignored. A umount from any one of those paths will result in the mount disappearing from both locations.

```shell
$ sudo umount /tmp/original/tmpfs 
vishnuk@vishnuk-linux:~$ cat /proc/self/mountinfo
...
```

jingxu97 replied:
The current solution used here is checking the mount path and the shared mount path created by the rkt container. If the shared mount path contains the original mount path, this will be filtered out during GetMountRef. See comments https://github.com/kubernetes/kubernetes/pull/38124#issuecomment-265094787

jingxu97 commented:
@k8s-bot cvm gke e2e test this

jingxu97 commented:
it is our own logic to check the mount references.

On Mon, Dec 5, 2016 at 10:52 PM, Vish Kannan <notifications@github.com>
wrote:

> *@vishh* commented on this pull request.
> ------------------------------
>
> In cluster/gce/gci/mounter/mounter
> <https://github.com/kubernetes/kubernetes/pull/38124>:
>
> > @@ -31,7 +31,7 @@ function gc {
>      # Rkt pods end up creating new copies of mounts on the host. Hence it is ideal to clean them up right away.
>      attempt=0
>      until [ $attempt -ge 5 ]; do
> -	${RKT_BINARY} gc --grace-period=0s &> /dev/null && break
>
> 1 might be of concern. Do we have any data to back up that concern?
> I don't get 2. Why is umount failing if other shared recursive mounts
> exist? Is the umount syscall failing or is it our own custom logic?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/38124>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASSNxeJff85Qf5NZp7QRL53kd8TJ770Zks5rFQY8gaJpZM4LEjHE>
> .
>



-- 
- Jing

saad-ali commented:
Please finalize on this PR ASAP, cutoff for 1.5 is Tuesday.

vishh commented:
I'm not comfortable with this PR because of https://github.com/kubernetes/kubernetes/pull/38124#discussion_r90998550

jingxu97 commented:
I am wondering whether it is enough to only check the propagation type and
its value. Isn't it possible that some other mount points that also have
the same value? By checking the mount path strings (see the example below),
the shared mount path should contain the full original mount path. So it
looks like by checking the mount paths and peer group X value (shared:X) it
might good enough to filter out the propagated mount refs.

351 314 8:16 / /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/nfs-disk
rw,relatime shared:133 - ext4 /dev/sdb rw,data=ordered
353 28 8:16 / /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/nfs-disk
rw,relatime shared:133 - ext4 /dev/sdb rw,data=ordered
352 911 8:16 /
/var/lib/rkt/pods/run/3496eadc-74d9-4349-9370-db7636f5f963/stage1/rootfs/opt/stage2/gci-mounter/rootfs/var/lib/kubelet/plugins/
kubernetes.io/gce-pd/mounts/nfs-disk rw,relatime shared:133 - ext4 /dev/sdb
rw,data=ordered


On Mon, Dec 5, 2016 at 11:22 PM, Vish Kannan <notifications@github.com>
wrote:

> *@vishh* commented on this pull request.
> ------------------------------
>
> In cluster/gce/gci/mounter/mounter
> <https://github.com/kubernetes/kubernetes/pull/38124>:
>
> > @@ -31,7 +31,7 @@ function gc {
>      # Rkt pods end up creating new copies of mounts on the host. Hence it is ideal to clean them up right away.
>      attempt=0
>      until [ $attempt -ge 5 ]; do
> -	${RKT_BINARY} gc --grace-period=0s &> /dev/null && break
>
> If a mount appears in two locations via a shared mount then
> /proc/self/mountinfo will explicitly indicate that. Here is an example.
>
> $ cat /proc/self/mountinfo
> ...
> 90 80 0:66 / /tmp/original/tmpfs rw,relatime **shared:2** - tmpfs tmpfs rw
> 103 101 0:66 / /tmp/shared/tmpfs rw,relatime **shared:2** - tmpfs tmpfs rw
>
> Look at the mount propagation type followed by a propagation number. If
> the number is the same, then that mount can be ignored. A umount from any
> one of those paths will result in the mount disappearing from both
> locations.
>
> $ sudo umount /tmp/original/tmpfs
> vishnuk@vishnuk-linux:~$ cat /proc/self/mountinfo
> ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/38124>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASSNxToh8NuuUoTS4PrS756Yaok4yvNQks5rFQ0_gaJpZM4LEjHE>
> .
>



-- 
- Jing

jingxu97 commented:
Updated the PR. PTAL.

saad-ali commented:
@vishh Without this change when anyone uses GlusterFS via mounter, all other mounts will fail to unmount. That's pretty bad. We need this fixed for the 1.5 release ASAP. This is hacky and I'd like to have a cleaner solution, but it gets the job done. @jingxu97 please open an issue to revisit and fix this code with a more robust solution.

saad-ali commented:
/lgtm

vishh commented:
I hope by merging this in a hurry, we did not introduce new bugs :)

saad-ali commented:
+1 We'll let it bake overnight in master. If the world hasn't burned down, we'll cherry pick this in the AM (Wed morning PST).

vishh commented:
It would help if Jing can add more comments to the PR explaining the logic
that she introduced using some example.

On Wed, Dec 7, 2016 at 8:26 AM, Saad Ali <notifications@github.com> wrote:

> +1 We'll let it bake overnight in master. If the world hasn't burned down,
> we'll cherry pick this in the AM (Wed morning PST).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/38124#issuecomment-265341465>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKJCXZ5sdN_UFxUYVWyqwkPMriI1Vks5rFiB6gaJpZM4LEjHE>
> .
>
